src/python/pants/option/options_fingerprinter.py

```diff
@@ -84,6 +79,11 @@ def fingerprint(self, option_type, option_val):
     if option_val is None:
       return None
 
+    # The python documentation (https://docs.python.org/2/library/json.html#json.dumps) states that
+    # dict keys are coerced to strings in json.dumps, but this appears to be incorrect.
+    if isinstance(option_val, dict):
+      option_val = {str(k):v for k, v in option_val.items()}
```

Comment:
If the key is not primitive, hashes are now subject to the whims of a potentially changing `__str__` impl. Seems better to leave this back up in the `Encoder` and if `encode` on a key - that's good, it means its not a primitive and we don't know how to stably encode it.

Reply:
Ok, that makes sense, I'll go back to having the encoder handle it.

Reply:
I wasn't able to figure out how to do the automatic encoding of dict keys, because if they're contained inside a list, the default json encoding will take over (and raise a `TypeError`), unless you either (1) special-case a list with a single dict in it (2) override the private method used to encode dicts (3) re-implement the whole of the `json` module. So instead I just made sure to hash (with `stable_json_sha1()`) the keys of all dicts we receive as `option_val` in the `fingerprint()` method. Does that sound reasonable, or is there a route I'm missing?

Reply:
Broke this out into #6422.